### PR TITLE
[VPlan] Only use selectVectorizationFactor for cross-check (NFCI).

### PR DESCRIPTION
**Pull request by fhahn**
Use getBestVF to select VF up-front and only use
selectVectorizationFactor to get the VF legacy VF to check the vectorization decision matches the VPlan-based cost model.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 1]
```diff
@@ -6995,11 +6994,11 @@ LoopVectorizationPlanner::plan(ElementCount UserVF, unsigned UserIC) {
...
         LLVM_DEBUG(dbgs() << "LV: No VPlan could be built for " << UserVF
                           << ".\n");
-        return std::nullopt;
+        return;
       }
 
       LLVM_DEBUG(printPlans(dbgs()));
-      return {{UserVF, 0, 0}};
+      return;
     } else
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:7002]: else can be dropped here

**Review by ayalz — COMMENTED** [review]
Good clean up of multiple MaybeVF, VF, BestVF!
Adding various comments.

**Review by ayalz — APPROVED** [review]
This LGTM, thanks for accommodating! Adding some thoughts for follow-up improvements.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 2]
```diff
@@ -7300,7 +7283,20 @@ ElementCount LoopVectorizationPlanner::computeBestVF() {
...
+#ifndef NDEBUG
+  // Select the optimal vectorization factor according to the legacy cost-model.
+  // This is now only used to verify the decisions by the new VPlan-based
+  // cost-model and will be retired once the VPlan-based cost-model is
+  // stabilized.
+  VectorizationFactor LegacyVF = selectVectorizationFactor();
+  assert(BestFactor.Width == LegacyVF.Width &&
+         " VPlan cost model and legacy cost model disagreed");
+  assert((BestFactor.Width.isScalar() || BestFactor.ScalarCost > 0) &&
+         "when vectorizing, the scalar cost must be non-zero.");
```
[the remark this answers is not in the document]
Comment (ayalz) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: As zero and/or max seem ambiguous - used as both artificial initial uncomputed values ("// A default constructed InstructionCost is a valid zero cost") and possibly considered valid computed costs, perhaps worth augmenting InstructionCost's `enum CostState` to include `UnSet` in addition to `Valid` and `InValid`?

One alternative could be to represent uncomputed costs via extreme out-of-bound values - getMinValue()-1 and/or getMaxValue()+1, somewhat analogous to -infinity and +infinity of floating-point numbers, where NaN represents InValid.

**Pull request merged**